### PR TITLE
ZIP 239 preparations 4

### PR DESCRIPTION
**Pull request by str4d**
Cherry-picked from the following upstream PRs:
- bitcoin/bitcoin#5913
  - Replaces zcash/zcash#3111.
  - Includes an extra commit included by upstream in the merge outside the PR.
- bitcoin/bitcoin#6519
- bitcoin/bitcoin#7179
- bitcoin/bitcoin#7853
- bitcoin/bitcoin#7960

### CODE REVIEW

**Comment by str4d** [comment]
@zkbot try

**Comment by str4d** [comment]
[Fixed the failing tests](https://github.com/zcash/zcash/compare/7a29d40a818fd2f2835ace1cc21577549d0a799a..1e352c108c0d2921c34854d2bbb21aa8af88f04d). Re-running to confirm.

@zkbot try

File: src/main.cpp [hunk 1]
```diff
@@ -2415,7 +2415,8 @@ void static InvalidBlockFound(CBlockIndex *pindex, const CValidationState &state
     if (state.IsInvalid(nDoS)) {
         std::map<uint256, NodeId>::iterator it = mapBlockSource.find(pindex->GetBlockHash());
         if (it != mapBlockSource.end() && State(it->second)) {
-            CBlockReject reject = {state.GetRejectCode(), state.GetRejectReason().substr(0, MAX_REJECT_MESSAGE_LENGTH), pindex->GetBlockHash()};
+            assert(state.GetRejectCode() < 0x100);
```
Comment (nuttycom) [inline src/main.cpp]: Is this the "extra commit included by upstream in the merge outside the PR"?

**Review by nuttycom — APPROVED** [review]
utACK verified against upstream commits.

File: src/main.cpp (hunk 1 above)
Comment (str4d) [reply src/main.cpp]: Yes. Upstream's GitHub shows it occuring in an unrelated PR, because it was not simply pushed to the original PR.

**Comment by str4d** [comment]
@zkbot r+

**Pull request merged**